### ISSUE TITLE
[CRITICAL] backend_hip.cpp stack buffer overflow: float logits[1000] receives VOCAB=262272 entries from zaya_forward

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#77]
## Location
`src/backend_hip.cpp` — `HIPBackend::forward()`, `HIPBackend::generate()`, `HIPBackend::benchmark()` (lines ~58, 79, 93)

## Description

The HIP backend declares a 1000-float stack buffer and passes it to `zaya_forward()`, which writes the full vocabulary logits into it. The actual vocabulary size is **262272**:

```cpp
// src/zaya_engine.cpp:21
static constexpr int N_LAYERS=40, VOCAB=262272, N_EXP=16, N_EXP_T=17, N_FF=2048, RTR_H=256;
```

But the HIP backend uses:
```cpp
// src/backend_hip.cpp
bool forward(int token_id, float* hidden_out) override {
    ...
    float logits[1000]; // small buffer        ← line 58
    zaya_forward(zs, token_id, logits);         ← writes 262272 floats (~1 MB)
    ...
    return false;  // TODO: wire hidden state output
}

int generate(int token_id) override {
    ...
    float logits[1000];                          ← line 79
    zaya_forward(zs, token_id, logits);          ← writes 262272 floats (~1 MB)
    int best = 0;
    for (int i = 1; i < 1000; i++) if (logits[i] > logits[best]) best = i;
    return best;
}

float benchmark(int tokens = 10) override {
    ...
        float logits[1000];                      ← line 93
        zaya_forward(zs, tok, logits);
    ...
}
```

`zaya_forward()` writes `VOCAB` (262272) `float`/`__half` entries into the caller's buffer regardless of its size. 262272 − 1000 = 261272 floats = **~1 MB of stack overwritten** past the end of `logits`.

## Impact

- **Stack buffer overflow on every call** to `forward()`, `generate()`, or `benchmark()` on the HIP backend. This smashes the stack frame — overwriting locals (`pos`, `best`, `tok`, `v`), saved registers, and ultimately the return address.
- Observable outcomes: immediate SIGSEGV/SIGBUS, corrupted argmax results (the `for (i = 1; i < 1000; i++)` scan only considers 1000 of 262272 vocab entries anyway, so even without a crash the sampled token is wrong), or exploitable control-flow hijack if an attacker can influence token weights/inputs.
- This is reachable via `BackendManager` (which loads the HIP backend through `create_hip_backend` / `librocm_cpp.so`) any time the HIP backend is selected and `generate()`/`forward()`/`benchmark_all()` is invoked.

## Suggested fix direction

- Size the buffer to `VOCAB` (262272), e.g. `float logits[VOCAB];` on the stack (≈1 MB — fine on most thread stacks but verify stack size) or, better, `std::vector<float> logits(VOCAB);` / a reused member buffer.
- Make the `argmax` scan range `VOCAB`, not the hardcoded `1000`.
- The `forward()` TODO (returns false without emitting hidden state, see separate issue on the HIP/NPU backend stubs) should be resolved in the same pass.

## Related
Distinct from #59 / #63 (those are about static-local device pointers and leaks inside `zaya_engine.cpp` itself). This is a host-side stack overflow in the `Backend`-adapter wrapper around that engine.

**Comment by bong-water-water-bong** [comment]
Fixed in `3bc3ff5` (pushed to main).

All three call sites (`forward`, `generate`, `benchmark`) now write into the already-allocated VOCAB-sized `logits_buf` member instead of a 1000-float stack buffer, and the argmax loops scan the full `VOCAB` (262272) rather than the first 1000 entries.

```diff
-        float logits[1000]; // small buffer
-        zaya_forward(zs, token_id, logits);
+        // Use the VOCAB-sized logits_buf member (allocated in init())
+        zaya_forward(zs, token_id, logits_buf);
 ...
-        for (int i = 1; i < 1000; i++) if (logits[i] > logits[best]) best = i;
+        for (int i = 1; i < VOCAB; i++) if (logits_buf[i] > logits_buf[best]) best = i;
```

Left open for you to verify on hardware / in a build; feel free to close once confirmed.
